### ISSUE TITLE
[interpreter] ruby - removal of symbols to the internal state/stack

### ISSUE THREAD

**Issue by florianl** [open-telemetry/opentelemetry-ebpf-profiler#202]
https://github.com/ruby/ruby/pull/7459 removed the access to symbols, that provided access to investigate the Ruby internal stack and enabled walking this interpeter stack.

https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/2c2cb6b9760350b0b9714c62a05d236791a276de/interpreter/ruby/ruby.go#L838-L844

Identify a new way to gain access to the Ruby internal state and stack to support newer versions of Ruby.

**Comment by dalehamel** [comment]
@florianl hey there! came across this and decided to file https://bugs.ruby-lang.org/issues/21539 as we'd love to see support for this. Do you have a list of symbols that otel unwinder depends on / would be good for ruby to export and keep stable?

**Comment by k0kubun** [comment]
My cursory reading of the ruby.go file tells me that it uses the symbols of `ruby_single_main_ractor` (or something that will give the current `ec`), `rb_vm_exec`, and `ruby_version`. If that's all it takes, it doesn't seem like a hard ask to keep such symbols (or something equivalent when changed) published.

**Comment by florianl** [comment]
sorry for the delayed response 🙏 
as @k0kubun pointed out, there are two main point this project is interested in, to support ruby unwinding.
Some kine of version indicator, like `ruby_version` and some pointer to the ruby VM, like `ruby_current_vm_ptr`, `ruby_current_vm` or `rb_vm_exec`.

Maybe also related on this topic is https://bugs.ruby-lang.org/issues/19119.

**Comment by dalehamel** [comment]
@florianl kind of inspired by the issue you referenced, i mocked up a means of exposing more minimal type information via an ELF section.

EDIT: Though, we can probably just do something similar using golang's debug/elf and debug/dwarf packages to parse this information directly off the binary rather than relying on the ruby folks to include it as part of the build?

I looked at https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/078ae4d6ded761b513038440bc8525014fa6c016/tools/coredump/testsources/ruby/gdb-dump-offsets.py#L30

And was able to get most of the information there at build time. For a PoC, i had just a header and a super minimal binary:

```stubs.h

#include "vm_core.h"
#include "ractor_core.h"
#include "method.h"
#include "iseq.h"

typedef struct rb_iseq_constant_body rb_iseq_constant_body_t;
typedef struct iseq_insn_info_entry iseq_insn_info_entry_t;

#include "include/ruby/internal/core/rstring.h"

typedef struct RString RString_t;

#include "include/ruby/internal/core/rarray.h"

typedef struct RArray RArray_t;

#define OFFSET_INFO(name, field) { #name "." #field, (size_t)offsetof(name, field), sizeof(((name *)0)->field) }
#define SIZE_INFO(name) { #name ".size_of_" #name, 0, sizeof(name) }

struct offset_info_entry {
    char name[128]; // NOTE - must ensure that the struct name and field name does not exceed this 128 chars.
    size_t offset;
    size_t size;
};

struct offset_info_entry offset_info[] __attribute__((section(".offset_info"))) = {
    OFFSET_INFO(rb_execution_context_t, vm_stack),
    OFFSET_INFO(rb_execution_context_t, vm_stack_size),
    OFFSET_INFO(rb_execution_context_t, cfp),

    OFFSET_INFO(rb_control_frame_t, pc),
    OFFSET_INFO(rb_control_frame_t, iseq),
    OFFSET_INFO(rb_control_frame_t, ep),
    SIZE_INFO(rb_control_frame_t),

    OFFSET_INFO(rb_iseq_t, body),

    OFFSET_INFO(rb_iseq_constant_body_t, type),
    OFFSET_INFO(rb_iseq_constant_body_t, iseq_size),
    OFFSET_INFO(rb_iseq_constant_body_t, iseq_encoded),
    OFFSET_INFO(rb_iseq_constant_body_t, location),
    OFFSET_INFO(rb_iseq_constant_body_t, insns_info.body),
    OFFSET_INFO(rb_iseq_constant_body_t, insns_info.size),
    OFFSET_INFO(rb_iseq_constant_body_t, insns_info.succ_index_table),
    SIZE_INFO(rb_iseq_constant_body_t),

    OFFSET_INFO(rb_iseq_location_t, pathobj),
    OFFSET_INFO(rb_iseq_location_t, base_label),

    OFFSET_INFO(iseq_insn_info_entry_t, line_no),
    SIZE_INFO(iseq_insn_info_entry_t),

    OFFSET_INFO(RString_t, as.embed.ary),
    OFFSET_INFO(RString_t, as.heap.ptr),

    OFFSET_INFO(RArray_t, as.ary),
    OFFSET_INFO(RArray_t, as.heap.ptr),

    // succ_dict_block and succ_index_table only defined in iseq.c, not able to reference it
    //'succ_index_table_struct.small_block_ranks': offset_of('succ_dict_block', 'small_block_ranks'),
    //'succ_index_table_struct.block_bits': offset_of('succ_dict_block', 'bits'),
    //'succ_index_table_struct.succ_part': offset_of('succ_index_table', 'succ_part'),
    //'succ_index_table_struct.size_of_succ_dict_block': size_of('succ_dict_block'),

    // Not sure what is up with this multiplication hack
    // 'size_of_immediate_table': size_of_field('succ_index_table', 'imm_part') * 9 // 8,

    SIZE_INFO(VALUE),

    OFFSET_INFO(rb_ractor_t, threads.running_ec),
};

```

```stubs.c
#include "stubs.h"

int main() {
    return 0;
}
```

(note that stubs.c is just to stand in lieu of the actual ruby binary, adding stubs.h to the ruby build should suffice to have the elf section added).

Compiling it with:

```
gcc -o stubs -I. -I./include -I.. -I./build/.ext/include/aarch64-linux -I./build stubs.c
```

I can read this data with:

```
objdump -s -j .offset_info stubs
```

So external profiles only need to know the structure of the offset_info entries in order to determine the offset and size information. This should help to avoid hardcoding things or trying to embed version specific headers. Eg,:

<img width="567" height="152" alt="Image" src="https://github.com/user-attachments/assets/3c3ccbb8-c68d-4dc6-a186-e4c3c18b74f3" />

But there are some drawbacks, as if the referenced structs change or we need more information to walk and symbolize the stack, this has to be kept up-to-date by the ruby maintainers. Also currently cannot read some fields as they are only defined in iseq.c

**Comment by florianl** [comment]
Thanks for the insights. The challenge from our external point of view is, that we don't have headers and don't compile at the endpoint. With the removal of public symbols by the upstream project, external profiler (but also debuggers like gdb, i think?) now have a more difficult time.

**Comment by dalehamel** [comment]
@florianl I have put together a draft for getting the field information from https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/707, and my colleague has some pending work on getting the symbols needed from the symbol table - i think with this we should have everything we need to be able to walk the stack on arbitrary ruby versions, but i still need to run the integration tests and ensure it is working end-to-end with the Loader

**Comment by florianl** [comment]
I agree with @fabled in https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/707#issuecomment-3199282272 for the named reasons and I think that we should not continue with https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/707.

As suggested, by @fabled we should either identify other symbols or metadata to get to the same information, ask upstream to make them available or disassemble code to extract needed information.

**Comment by manuelfelipe** [comment]
> As suggested, by @fabled we should either identify other symbols or metadata to get to the same information, ask upstream to make them available or disassemble code to extract needed information.


@florianl, would something like https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/710 be fine in the meantime? Unsure if falling back to the full symtab for discovery is also not ideal. Went that route because of existing presence in golabels but happy to discuss and explore alternatives.
